hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/BackgroundSCMService.java

```diff
@@ -0,0 +1,208 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.ratis.util.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Clock;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A common implementation for background SCMService.
+ * */
+public final class BackgroundSCMService implements SCMService {
+  private final Logger log;
+  private final AtomicBoolean running = new AtomicBoolean(false);
+  private Thread backgroundThread;
+  private final Lock serviceLock = new ReentrantLock();
+  private final SCMContext scmContext;
+  private ServiceStatus serviceStatus = ServiceStatus.PAUSING;
+  private final long intervalInMillis;
+  private final long waitTimeInMillis;
+  private long lastTimeToBeReadyInMillis = 0;
+  private final Clock clock;
+  private String serviceName;
+  private Runnable periodicalTask;
+  private volatile boolean runImmediately = false;
+
+  private BackgroundSCMService(
+      final Clock clock, final SCMContext scmContext,
+      final String serviceName, final long intervalInMillis,
+      final long waitTimeInMillis, final Runnable task) {
+    this.scmContext = scmContext;
+    this.clock = clock;
+    this.periodicalTask = task;
+    this.serviceName = serviceName;
+    this.log = LoggerFactory.getLogger(serviceName);
+    this.intervalInMillis = intervalInMillis;
+    this.waitTimeInMillis = waitTimeInMillis;
+    start();
+  }
+
+  @Override
+  public void start() {
+    if (!running.compareAndSet(false, true)) {
+      log.info("{} Service is already running, skip start.", getServiceName());
+      return;
+    }
+    log.info("Starting {} Service.", getServiceName());
+
+    backgroundThread = new Thread(this::run);
+    backgroundThread.setName(serviceName + "Thread");
+    backgroundThread.setDaemon(true);
+    backgroundThread.start();
+  }
+
+  @Override
+  public void notifyStatusChanged() {
+    serviceLock.lock();
+    try {
+      if (scmContext.isLeaderReady() && !scmContext.isInSafeMode()) {
+        if (serviceStatus != ServiceStatus.RUNNING) {
+          log.info("Service {} transitions to RUNNING.", getServiceName());
+          serviceStatus = ServiceStatus.RUNNING;
+          lastTimeToBeReadyInMillis = clock.millis();
+        }
+      } else {
+        if (serviceStatus != ServiceStatus.PAUSING) {
+          log.info("Service {} transitions to PAUSING.", getServiceName());
+          serviceStatus = ServiceStatus.PAUSING;
+        }
+      }
+    } finally {
+      serviceLock.unlock();
+    }
+  }
+
+  private void run() {
+    while (running.get()) {
+      try {
+        if (shouldRun()) {
+          periodicalTask.run();
+        }
+        synchronized (this) {
+          if (!runImmediately) {
+            wait(intervalInMillis);
+          }
+          runImmediately = false;
+        }
+      } catch (InterruptedException e) {
```

Comment:
Should we catch and perhaps log any exceptions here? I've seen cases where these sort of background threads get an error and then just exit with no logging at all. I think we should at least log the error, and consider letting the job retry next time.

However the way the code is structured now, if we get an error in `periodicalTask.run()` and catch it to let it retry, it will not wait and will retry over and over immediately. We probably don't want that either!

Reply:
Also I am not sure if a runnable is allowed to throw checked exceptions? Can it only throw RuntimeExceptions? Maybe its enough to catch and log RuntimeException and then re-throw? I am not sure.

Reply:
`Runnable` can't throw a checked exception because it is not declared to do so. we can't throw checked exceptions without declaring them. if there is any exception in `run`, it should be caught and handled(Eg. log this error) inside itself , but not be thrown out.
so , i suggest we keep the code like this and implement error or exception handling logic inside `periodicalTask.run()` itself